### PR TITLE
Add protected handlers for next

### LINKED ISSUE #94: Create Webhook Handler factory
`createSaleorWebhookHandler` exported from `@saleor/app-sdk/handler/next`

```
// src/pages/api/webhooks/order_created.ts

export default createSaleorWebhookHandler({
  eventType: string | string[] 
  ...
})

```

```
// eg:
const handler: SaleorEventHandler<OrderCreatedDocument | OrderFullfilledDocument> = (req, res) => {
  const {apl, payload, domain, eventType} = req.context;

  // logic
}

export default createSaleorWebhookHandler({
  handler: handler
})
```

- [x] Check what Saleor event type
- [x] Inject SaleorApp / APL instance
- [x] Validate tokens and headers
- [x] Check payload checksum
- [x] Type extending Next's request with Saleors metadata, eg `SaleorEventRequest<PayloadGraphqlDocument>`
- [ ] Allow multiple event types
- [x] Documentation
- [x] Example

### LINKED ISSUE #104: Create factory for handlers available only from the dashboard
Applications usually have endpoints to fetch/update app configuration used by the App interface inside of the Dashboard.

The factory should return handler validating incoming requests and their tokens.

Ref #94
Ref https://github.com/saleor/react-storefront/pull/592

### PR DESCRIPTION
**Pull request by krzysztofwolski**
Close #104 

Usage example can be found here: https://github.com/saleor/saleor-app-template/pull/119

### CODE REVIEW

File: src/handlers/next/create-protected-handler.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+  res: NextApiResponse<TResp>,
+  ctx: ProtectedHandlerContext
+) => unknown | Promise<unknown>;
+
+/**
+ * Wraps provided function, to ensure incoming request comes from Saleor Dashboard.
+ * Also provides additional `context` object containing request properties.
+ *
+ * @param handlerFn NextApiHandler function which takes additional `context` argument
+ * @returns NextApiHandler
```
Comment (lkostrowski) [inline src/handlers/next/create-protected-handler.ts:35]: Lets remove autogenerated param jsdoc statements, they are redundant in typescript. Type NextApiHandler can be described with extra comment

File: src/handlers/next/create-protected-handler.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+ */
+export const createProtectedHandler =
+  (handlerFn: NextProtectedApiHandler, apl: APL): NextApiHandler =>
+  async (req, res) => {
+    debug("Protected handler called");
+    await processSaleorProtectedHandler({
```
Comment (lkostrowski) [inline src/handlers/next/create-protected-handler.ts:41]: you are using async/await + then/catch. Maybe lets stick to one for simplicity?

File: src/handlers/next/process-protected-handler.ts [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    throw new ProtectedHandlerError("Missing saleor-domain header", "MISSING_DOMAIN_HEADER");
+  }
+
+  if (!token) {
+    debug("Missing authorization-bearer header");
+    throw new ProtectedHandlerError(
+      "Missing authorization-bearer header",
+      "MISSING_AUTHORIZATION_BEARER_HEADER"
+    );
+  }
```
Comment (lkostrowski) [inline src/handlers/next/process-protected-handler.ts:75]: Not scope of this PR but maybe we should write some common assertions library, that will share this logic, eg

```
assert.authHeader(req); // void or throw
```

We will have unified errors which are technically the same, we can add context of which handler in extra argument

**Review by lkostrowski — APPROVED** [review]

File: src/handlers/next/create-protected-handler.ts (hunk 2 above)
Comment (krzysztofwolski) [reply src/handlers/next/create-protected-handler.ts:41]: Turns out it was not needed. Removed await

File: src/handlers/next/create-protected-handler.ts (hunk 1 above)
Comment (krzysztofwolski) [reply src/handlers/next/create-protected-handler.ts:35]: Removed 👍

File: src/handlers/next/process-protected-handler.ts (hunk 3 above)
Comment (krzysztofwolski) [reply src/handlers/next/process-protected-handler.ts:75]: Love the idea. Added issue #125

**Review by lkostrowski — APPROVED** [review]

**Pull request merged**